### ISSUE TITLE
What is the correct format for --pulse-device-name?

### ISSUE THREAD

**Issue by vrrobz** [ideasman42/nerd-dictation#20]
First off - thank you. This is precisely what I have been looking for. Great work here!

I want to ensure that the program is using the right microphone - I want to make sure it uses the external one, not the one on my laptop. Running `pactl list` gives me a WHOLE slew of stuff, but I think this is the chunk I'm most interested in, since it lists my external microphone:

```
Card #2
	Name: alsa_card.usb-BLUE_MICROPHONE_Blue_Snowball_201603-00
	Driver: module-alsa-card.c
	Owner Module: 28
	Properties:
		alsa.card = "1"
		alsa.card_name = "Blue Snowball"
		alsa.long_card_name = "BLUE MICROPHONE Blue Snowball at usb-0000:00:14.0-3, full speed"
		alsa.driver_name = "snd_usb_audio"
		device.bus_path = "pci-0000:00:14.0-usb-0:3:1.0"
		sysfs.path = "/devices/pci0000:00/0000:00:14.0/usb1/1-3/1-3:1.0/sound/card1"
		udev.id = "usb-BLUE_MICROPHONE_Blue_Snowball_201603-00"
		device.bus = "usb"
		device.vendor.id = "0d8c"
		device.vendor.name = "C-Media Electronics, Inc."
		device.product.id = "0005"
		device.product.name = "Blue Snowball"
		device.serial = "BLUE_MICROPHONE_Blue_Snowball_201603"
		device.string = "1"
		device.description = "Blue Snowball"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-usb"
	Profiles:
		input:mono-fallback: Mono Input (sinks: 0, sources: 1, priority: 1, available: yes)
		input:multichannel-input: Multichannel Input (sinks: 0, sources: 1, priority: 1, available: yes)
		off: Off (sinks: 0, sources: 0, priority: 0, available: yes)
	Active Profile: input:mono-fallback
	Ports:
		analog-input-mic: Microphone (priority: 8700, latency offset: 0 usec)
			Properties:
				device.icon_name = "audio-input-microphone"
			Part of profile(s): input:mono-fallback
		multichannel-input: Multichannel Input (priority: 0, latency offset: 0 usec)
			Part of profile(s): input:multichannel-input
```

I have tried feeding the "Name" value (alsa_card.usb-BLUE_MICROPHONE_Blue_Snowball_201603-00), the udev.id, and the device.icon_name (longshot) into the CLI, each time getting the error `Stream error: No such entity`. If I don't include the `--pulse-device-name`, dictation works fine, but I want to ensure it's getting the best input possible. 

Which of the values from the `pactl list` output should we use for that flag? Or is there another value further up in the stream - i.e. not "Card #2' - that I should be looking at?

Thanks!

**Comment by ideasman42** [comment]
Hi, the identifier following `Name: ` should be sufficient.

Although it looks like you've tried this and it's not working for you.

Your best bet for trouble shooting this problem is to run `parec` directly to figure out the command needed to record from your audio device.

I use a wrapper like this: to begin dictation`nerd-dictation-begin`
```
#!/bin/sh
exec \
    nerd-dictation \
    begin \
    --vosk-model-dir=/path-to/src/vosk-api/python/example/model \
    --pulse-device-name=alsa_input.usb-M-Audio_AIR_192_4-00.analog-stereo \
    --delay-exit=0.1 \
    --numbers-as-digits \
    --numbers-use-separator \
    "$@"
```

`pactl list` shows this output (many others too, but this is the device I use for recording).

```
Source #1
	State: SUSPENDED
	Name: alsa_input.usb-M-Audio_AIR_192_4-00.analog-stereo
	Description: AIR 192 4 Analog Stereo
	Driver: module-alsa-card.c
	Sample Specification: s32le 2ch 44100Hz
	Channel Map: front-left,front-right
	Owner Module: 8
	Mute: no
	Volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
	        balance 0.00
	Base Volume: 65536 / 100% / 0.00 dB
	Monitor of Sink: n/a
	Latency: 0 usec, configured 0 usec
	Flags: HARDWARE DECIBEL_VOLUME LATENCY 
	Properties:
		alsa.resolution_bits = "32"
		device.api = "alsa"
		device.class = "sound"
		alsa.class = "generic"
		alsa.subclass = "generic-mix"
		alsa.name = "USB Audio"
		alsa.id = "USB Audio"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.device = "0"
		alsa.card = "3"
		alsa.card_name = "AIR 192 4"
		alsa.long_card_name = "M-Audio AIR 192 4 at usb-0000:25:00.3-2, high speed"
		alsa.driver_name = "snd_usb_audio"
		device.bus_path = "pci-0000:25:00.3-usb-0:2:1.0"
		sysfs.path = "/devices/pci0000:20/0000:20:08.1/0000:25:00.3/usb3/3-2/3-2:1.0/sound/card3"
		udev.id = "usb-M-Audio_AIR_192_4-00"
		device.bus = "usb"
		device.vendor.id = "0763"
		device.vendor.name = "M-Audio"
		device.product.id = "410b"
		device.product.name = "AIR 192 4"
		device.serial = "M-Audio_AIR_192_4"
		device.string = "front:3"
		device.buffering.buffer_size = "705600"
		device.buffering.fragment_size = "352800"
		device.access_mode = "mmap+timer"
		device.profile.name = "analog-stereo"
		device.profile.description = "Analog Stereo"
		device.description = "AIR 192 4 Analog Stereo"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-usb"
	Ports:
		analog-input: Analog Input (type: Analog, priority: 10000, availability unknown)
	Active Port: analog-input
	Formats:
		pcm
```

To check if you can record a `.wav` file with parec, try something like this:

`parec --format=s16le --rate=44100 --device=alsa_input.usb-M-Audio_AIR_192_4-00.analog-stereo --file-format=wav test.wav`

**Comment by zjuyk** [comment]
It depens on your pulseaudio config. For example, I use pulseaudio-alsa, then I will get the device name via

```
$ pactl list | grep -A2 'Source #' | grep 'Name: ' | cut -d" " -f2
alsa_output.pci-0000_00_1f.3.analog-stereo.monitor
alsa_input.pci-0000_00_1f.3.analog-stereo
```
the output and input name is what you want

**Comment by ideasman42** [comment]
@vrrobz did you manage to resolve this issue?

**Comment by vrrobz** [comment]
I apologize for not following up earlier.

The `pactl list` command returns a WHOLE BUNCH of options. I guess I was
choosing the wrong one. Following your example above, I looked under the
"Source" items for a mention of the Blue Snowball mic. I grabbed the "name"
field from that and fed it into the `--pulse-device-name` property when I
ran `nerd-dictation`. Worked like a charm.

Perhaps the docs should be updated to suggest looking for the right device
as one of the "Sources".  I was using the values under the "Card" section -
they don't match.

```
Card #7
Name: alsa_card.usb-BLUE_MICROPHONE_Blue_Snowball_201603-00
...
```

```
Source #9
State: SUSPENDED
Name: alsa_input.usb-BLUE_MICROPHONE_Blue_Snowball_201603-00.mono-fallback
Description: Blue Snowball Mono
...
```

The "State: SUSPENDED" bit also threw me - it;s why I went with the Card
value instead of the Source value to begin with.

Thanks for your help and quick response!
